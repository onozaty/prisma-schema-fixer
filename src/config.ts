import { EnumMapRule } from "./rules/enum-map-rule";
import { EnumNameRule } from "./rules/enuml-name-rule";
import { FieldAttributeRule } from "./rules/field-attribute-rule";
import { FieldMapRule } from "./rules/field-map-rule";
import { FieldNameRule } from "./rules/field-name-rule";
import { ModelMapRule } from "./rules/model-map-rule";
import { ModelNameRule } from "./rules/model-name-rule";

export type Config = {
  rules: {
    "model-name"?: ModelNameRule.Config[];
    "model-map"?: ModelMapRule.Config[];
    "field-name"?: FieldNameRule.Config[];
    "field-map"?: FieldMapRule.Config[];
    "field-attribute"?: FieldAttributeRule.Config[];
    "enum-name"?: EnumNameRule.Config[];
    "enum-map"?: EnumMapRule.Config[];
  };
};

export const defineConfig = (config: Config): Config => {
  return config;
};
