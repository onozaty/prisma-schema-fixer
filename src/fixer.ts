import { Config } from "./config";
import { joinBlocks, parseBlocks } from "./parser";
import { EnumMapRule } from "./rules/enum-map-rule";
import { EnumNameRule } from "./rules/enuml-name-rule";
import { FieldMapRule } from "./rules/field-map-rule";
import { FieldNameRule } from "./rules/field-name-rule";
import { ModelMapRule } from "./rules/model-map-rule";
import { ModelNameRule } from "./rules/model-name-rule";
import { formatSchema } from "./schema";

export const fix = async (content: string, config: Config): Promise<string> => {
  const blocks = parseBlocks(await formatSchema(content));

  if (config.rules["model-name"] !== undefined) {
    const configs = config.rules["model-name"];
    ModelNameRule.apply(configs, blocks);
  }
  if (config.rules["model-map"] !== undefined) {
    const configs = config.rules["model-map"];
    ModelMapRule.apply(configs, blocks);
  }
  if (config.rules["field-name"] !== undefined) {
    const configs = config.rules["field-name"];
    FieldNameRule.apply(configs, blocks);
  }
  if (config.rules["field-map"] !== undefined) {
    const configs = config.rules["field-map"];
    FieldMapRule.apply(configs, blocks);
  }
  if (config.rules["enum-name"] !== undefined) {
    const configs = config.rules["enum-name"];
    EnumNameRule.apply(configs, blocks);
  }
  if (config.rules["enum-map"] !== undefined) {
    const configs = config.rules["enum-map"];
    EnumMapRule.apply(configs, blocks);
  }

  const fixedContent = joinBlocks(blocks);
  return formatSchema(fixedContent);
};
