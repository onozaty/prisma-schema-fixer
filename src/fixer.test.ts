import fs from "fs";
import path from "path";
import { defineConfig } from "./config";
import { fix } from "./fixer";

describe("fix", () => {
  test("model-name", async () => {
    // Arrange
    const content = readFixture("simple.prisma");
    const config = defineConfig({
      rules: {
        "model-name": [
          {
            case: "pascal",
            form: "singular",
          },
        ],
      },
    });

    // Act
    const result = await fix(content, config);

    // Assert
    expect(result).toMatchSnapshot();
  });
  test("model-map", async () => {
    // Arrange
    const content = readFixture("simple.prisma");
    const config = defineConfig({
      rules: {
        "model-map": [
          {
            case: "snake",
            form: "plural",
          },
        ],
      },
    });

    // Act
    const result = await fix(content, config);

    // Assert
    expect(result).toMatchSnapshot();
  });
  test("field-name", async () => {
    // Arrange
    const content = readFixture("simple.prisma");
    const config = defineConfig({
      rules: {
        "field-name": [
          {
            case: "pascal",
          },
        ],
      },
    });

    // Act
    const result = await fix(content, config);

    // Assert
    expect(result).toMatchSnapshot();
  });
  test("field-map", async () => {
    // Arrange
    const content = readFixture("simple.prisma");
    const config = defineConfig({
      rules: {
        "field-map": [
          {
            case: "snake",
          },
        ],
      },
    });

    // Act
    const result = await fix(content, config);

    // Assert
    expect(result).toMatchSnapshot();
  });
  test("enum-name", async () => {
    // Arrange
    const content = readFixture("simple.prisma");
    const config = defineConfig({
      rules: {
        "enum-name": [
          {
            case: "pascal",
            form: "plural",
          },
        ],
      },
    });

    // Act
    const result = await fix(content, config);

    // Assert
    expect(result).toMatchSnapshot();
  });
  test("enum-map", async () => {
    // Arrange
    const content = readFixture("simple.prisma");
    const config = defineConfig({
      rules: {
        "enum-map": [
          {
            case: "snake",
            form: "singular",
          },
        ],
      },
    });

    // Act
    const result = await fix(content, config);

    // Assert
    expect(result).toMatchSnapshot();
  });
  test("all", async () => {
    // Arrange
    const content = readFixture("simple.prisma");
    const config = defineConfig({
      rules: {
        "model-name": [
          {
            case: "pascal",
            form: "singular",
          },
        ],
        "model-map": [
          {
            case: "snake",
            form: "plural",
          },
        ],
        "field-name": [
          {
            case: "camel",
          },
        ],
        "field-map": [
          {
            case: "snake",
          },
        ],
        "enum-name": [
          {
            case: "pascal",
            form: "singular",
          },
        ],
        "enum-map": [
          {
            case: "snake",
            form: "plural",
          },
        ],
      },
    });

    // Act
    const result = await fix(content, config);

    // Assert
    expect(result).toMatchSnapshot();
  });
});

const readFixture = (name: string): string => {
  const fixturePath = path.join(__dirname, "__fixtures__", name);
  return fs.readFileSync(fixturePath, "utf-8");
};
