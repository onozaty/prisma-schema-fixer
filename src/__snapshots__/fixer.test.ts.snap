// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`fix > all 1`] = `
"datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

generator client {
  provider = "prisma-client-js"
}

model User {
  id        Int          @id @default(autoincrement())
  createdAt DateTime     @default(now()) @map("created_at")
  email     String       @unique
  role      Role         @default(USER)
  posts     Post[]
  profile   UserProfile?

  @@map("users")
}

model UserProfile {
  id     Int     @id @default(autoincrement())
  name   String?
  bio    String?
  user   User    @relation(fields: [userId], references: [id])
  userId Int     @unique @map("user_id")

  @@map("user_profiles")
}

model Post {
  id        Int      @id @default(autoincrement())
  createdAt DateTime @default(now()) @map("created_at")
  updatedAt DateTime @updatedAt @map("updated_at")
  published Boolean  @default(false)
  title     String   @db.VarChar(255)
  author    User?    @relation(fields: [authorId], references: [id])
  authorId  Int?     @map("author_id")

  @@map("posts")
}

enum Role {
  USER
  ADMIN

  @@map("roles")
}
"
`;

exports[`fix > enum-map 1`] = `
"datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

generator client {
  provider = "prisma-client-js"
}

model Users {
  id        Int          @id @default(autoincrement())
  createdAt DateTime     @default(now())
  email     String       @unique
  role      Role         @default(USER)
  posts     Post[]
  profile   UserProfile?
}

model userProfile {
  id     Int     @id @default(autoincrement())
  name   String?
  bio    String?
  user   User    @relation(fields: [userId], references: [id])
  userId Int     @unique
}

model post {
  Id        Int      @id @default(autoincrement())
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt
  published Boolean  @default(false)
  title     String   @db.VarChar(255)
  author    User?    @relation(fields: [author_id], references: [id])
  author_id Int?
}

enum Role {
  USER
  ADMIN

  @@map("role")
}
"
`;

exports[`fix > enum-name 1`] = `
"datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

generator client {
  provider = "prisma-client-js"
}

model Users {
  id        Int          @id @default(autoincrement())
  createdAt DateTime     @default(now())
  email     String       @unique
  role      Roles        @default(USER)
  posts     Post[]
  profile   UserProfile?
}

model userProfile {
  id     Int     @id @default(autoincrement())
  name   String?
  bio    String?
  user   User    @relation(fields: [userId], references: [id])
  userId Int     @unique
}

model post {
  Id        Int      @id @default(autoincrement())
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt
  published Boolean  @default(false)
  title     String   @db.VarChar(255)
  author    User?    @relation(fields: [author_id], references: [id])
  author_id Int?
}

enum Roles {
  USER
  ADMIN
}
"
`;

exports[`fix > field-map 1`] = `
"datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

generator client {
  provider = "prisma-client-js"
}

model Users {
  id        Int          @id @default(autoincrement())
  createdAt DateTime     @default(now()) @map("created_at")
  email     String       @unique
  role      Role         @default(USER)
  posts     Post[]
  profile   UserProfile?
}

model userProfile {
  id     Int     @id @default(autoincrement())
  name   String?
  bio    String?
  user   User    @relation(fields: [userId], references: [id])
  userId Int     @unique @map("user_id")
}

model post {
  Id        Int      @id @default(autoincrement()) @map("id")
  createdAt DateTime @default(now()) @map("created_at")
  updatedAt DateTime @updatedAt @map("updated_at")
  published Boolean  @default(false)
  title     String   @db.VarChar(255)
  author    User?    @relation(fields: [author_id], references: [id])
  author_id Int?
}

enum Role {
  USER
  ADMIN
}
"
`;

exports[`fix > field-name 1`] = `
"datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

generator client {
  provider = "prisma-client-js"
}

model Users {
  Id        Int          @id @default(autoincrement())
  CreatedAt DateTime     @default(now())
  Email     String       @unique
  Role      Role         @default(USER)
  Posts     Post[]
  Profile   UserProfile?
}

model userProfile {
  Id     Int     @id @default(autoincrement())
  Name   String?
  Bio    String?
  User   User    @relation(fields: [UserId], references: [id])
  UserId Int     @unique
}

model post {
  Id        Int      @id @default(autoincrement())
  CreatedAt DateTime @default(now())
  UpdatedAt DateTime @updatedAt
  Published Boolean  @default(false)
  Title     String   @db.VarChar(255)
  Author    User?    @relation(fields: [AuthorId], references: [id])
  AuthorId  Int?
}

enum Role {
  USER
  ADMIN
}
"
`;

exports[`fix > model-map 1`] = `
"datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

generator client {
  provider = "prisma-client-js"
}

model Users {
  id        Int          @id @default(autoincrement())
  createdAt DateTime     @default(now())
  email     String       @unique
  role      Role         @default(USER)
  posts     Post[]
  profile   UserProfile?

  @@map("users")
}

model userProfile {
  id     Int     @id @default(autoincrement())
  name   String?
  bio    String?
  user   User    @relation(fields: [userId], references: [id])
  userId Int     @unique

  @@map("user_profiles")
}

model post {
  Id        Int      @id @default(autoincrement())
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt
  published Boolean  @default(false)
  title     String   @db.VarChar(255)
  author    User?    @relation(fields: [author_id], references: [id])
  author_id Int?

  @@map("posts")
}

enum Role {
  USER
  ADMIN
}
"
`;

exports[`fix > model-name 1`] = `
"datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

generator client {
  provider = "prisma-client-js"
}

model User {
  id        Int          @id @default(autoincrement())
  createdAt DateTime     @default(now())
  email     String       @unique
  role      Role         @default(USER)
  posts     Post[]
  profile   UserProfile?
}

model UserProfile {
  id     Int     @id @default(autoincrement())
  name   String?
  bio    String?
  user   User    @relation(fields: [userId], references: [id])
  userId Int     @unique
}

model Post {
  Id        Int      @id @default(autoincrement())
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt
  published Boolean  @default(false)
  title     String   @db.VarChar(255)
  author    User?    @relation(fields: [author_id], references: [id])
  author_id Int?
}

enum Role {
  USER
  ADMIN
}
"
`;
